### ISSUE TITLE
[BUG] The require-memo rule does not work if the component is wrapped in memo not directly.

### ISSUE THREAD

**Issue by veksa** [arthurgeron/eslint-plugin-react-usememo#33]
## Minimal repro

Please provide a minimal piece of code or reproducible example that demonstrates the issue.  

```javascript
const Component = () => <div />; const ComponentExport = memo(Component)
``````   
> Some rules process import statements. Make sure to include them if they're relevant to this issue.

### Installed packages versions
**ESLint:**  8.52.0
**Eslint-plugin-react-usememo:** 2.2.1

> You can get these details by using `npm ls --depth=0`.

### Rule and its options

Which rule from the eslint-plugin-react-usememo is being used? What configuration options are set for this rule?
'@arthurgeron/react-usememo/require-usememo': [2, {
    'checkHookReturnObject': true,
    'checkHookCalls': true,
    'fix': {'addImports': false},
}],

**Comment by arthurgeron** [comment]
I've analyzed the logic for that rule and it seems like a slippery slope to check all components declarations regardless of context. So I've done the following changes:
- I've narrowed down the validation  to exports, default and named, so that we can precisely validate only what's necessary. 
- Stopped checking other declarations (e.g. components dynamically declared inside components), and there are a few reasons for this:
  -  It overlaps with the `require-usememo`'s validations.
  -  In AST we need a "point of entry" for our validations, we can use declarations, exports, anything you think of. On your specific case, we'd need to check every declaration and see if it's declared again with a wrapping `memo`, which is a huge pain and very finicky.



So, within those parameters, your issue has been fixed. I'd just like you to test and give me some feedback before I publish it. It's available on the beta `@arthurgeron/eslint-plugin-react-usememo@2.3.0--beta0`.

Thank you for reporting this bug!

**Comment by veksa** [comment]
checked, rule is crashed with error:

`TypeError: Cannot read properties of undefined (reading 'toUpperCase')

ruleId: '@arthurgeron/react-usememo/require-usememo',`

minimal code to reproduce:

`const test = (use: Function) => {
    use();
};`

**Comment by arthurgeron** [comment]
@veksa That's a new issue, the lib didn't support React Canary versions yet. No longer the case, added support for it in main and beta versions.
New beta: `@arthurgeron/eslint-plugin-react-usememo@2.3.0--beta1`

**Comment by veksa** [comment]
@arthurgeron The previous issue has been resolved, but in this case, a new error has emerged:

`export const variable = func()();`

`TypeError: Cannot read properties of undefined (reading 'name')
Rule: "@arthurgeron/react-usememo/require-memo"`

**Comment by arthurgeron-work** [comment]
Thanks @veksa, I'll take a look as soon as I can. Is the code you provided on your previous comment enough to reproduce that issue?

**Comment by veksa** [comment]
@arthurgeron yes, i've written the minimal code for reproduce.

**Comment by arthurgeron-work** [comment]
@veksa Please check again with the beta `@arthurgeron/eslint-plugin-react-usememo@2.3.0--beta`

**Comment by veksa** [comment]
@arthurgeron everything is now functioning without errors.

However, there is a rule triggering errors:

`Component definition not wrapped in React.memo()  @arthurgeron/react-usememo/require-memo`

for all exported uppercase constants :)

for example:

`export const TestMap = {};`

Perhaps consider checking if the interface is FC or FunctionComponent to determine whether it is a component instead of just a constant?

**Comment by arthurgeron-work** [comment]
Yeah that's not right, it should be checking if it's a a valid function component

**Comment by arthurgeron-work** [comment]
Now it won't try to validate expressions or objects that aren't within the expected group. Please check again with `@arthurgeron/eslint-plugin-react-usememo@2.3.0--beta3`, and thank you for the quick feedback.

**Comment by arthurgeron-work** [comment]
@veksa Now it won't try to validate expressions or objects that aren't within the expected group. Please check again with `@arthurgeron/eslint-plugin-react-usememo@2.3.0--beta3`, and thank you for the quick feedback.

**Comment by arthurgeron-work** [comment]
@veksa  I double checked with `npm view` and the 2.3.0--beta3 was published! 
<img width="1185" alt="Captura de Tela 2024-01-10 às 06 48 13" src="https://github.com/arthurgeron/eslint-plugin-react-usememo/assets/71399856/bd042a89-2651-4348-a3ff-0bd4af854f9e">

**Comment by veksa** [comment]
yep, maybe on my side. everything looks fine when i use `npm view`. however, when checked with `yarn npm info @arthurgeron/eslint-plugin-react-usememo`:

there is no `2.3.0--beta3`, only `2.3.0--beta2`.

anyway, I checked without it.

in most cases, everything worked correctly.

however, there is a case when it does not work correctly:

a simple function with an uppercase name and parameters:

`export function Test(params: any) {}` or `export const Test = (params: any) => {}`

it seems like it's too complex to check that here; I will mark it as ignored in my case. :)

**Suggestions**:

add an option to ignore paths specified by a regular expression (for example, components in stories that do not require memoization at all). This option will be more practical than the ignoreComponents option.

introduce an option to ignore objects/arrays without calculations in hook properties. in this case, useMemo will function identically in terms of memory usage. For example:
`useSelector(someSelector, {})`

consider implementing a mechanism to directly mark hooks for exclusion, as maintaining a list of ignored hooks in a complex project can be challenging. :)

**Comment by arthurgeron-work** [comment]
That's odd!  It could be some DNS cache issue on your side, yesterday, I was able to install the beta03 on one of my projects.
> however, there is a case when it does not work correctly:
a simple function with an uppercase name and parameters:
`export function Test(params: any) {} or export const Test = (params: any) => {}`
it seems like it's too complex to check that here; I will mark it as ignored in my case. :).  

Don't worry too much about complexity, AST is more straightforward than it seems. Checking if a component has parameters is a great idea and something I'll now include in the next beta, hold tight!

> add an option to ignore paths specified by a regular expression (for example, components in stories that do not require memoization at all). This option will be more practical than the ignoreComponents option.   

I don't think it's a rule's responsibility to manage folder exclusion; that can be done easily through eslint's [overrides](https://eslint.org/docs/latest/use/configure/rules#using-configuration-files-1), something similar has been discussed [here](https://github.com/arthurgeron/eslint-plugin-react-usememo/issues/26#issuecomment-1824884766) which can be used as an example; you can create an override for your stories folder and completely disable the plugin or specify different options. 

Also, I'm not a fan of using Regexp in ESLint rules, it's heavy and it scales badly, specifically if you consider how many functions/expressions it'll be checking at the same time when being run through an entire project. 

>introduce an option to ignore objects/arrays without calculations in hook properties. in this case, useMemo will function identically in terms of memory usage. For example:
useSelector(someSelector, {})   

This one got me a little confused, are we still talking about the same rule?   
When deciding what needs or not a useMemo/callback, rather than checking complexity we look at referential equality, there's a nice article about that [here](https://blog.bitsrc.io/understanding-referential-equality-in-react-a8fb3769be0#:~:text=React%20compares%20the%20new%20value,to%20reflect%20the%20state%20change.).   
The focus of requiring useMemo/useCallback for hook return statements, props, and hook calls is that we don't want React, hooks, or selectors to think that something has changed when it has not, which would, in turn, cause unnecessary rerenders, potentially exponentially, slowing down your application and causing unexpected behavior.

>consider implementing a mechanism to directly mark hooks for exclusion, as maintaining a list of ignored hooks in a complex project can be challenging. :)   

Instead of regex or just 1:1 comparison, I could make the "ignoreComponents" option support [minimatch](https://www.npmjs.com/package/minimatch) for comparisons, which supports glob expressions, it'd benefit projects using standards for components names. Something similar to what's already present in the `ignoredHookCallsNames` exclusion option for [require-usememo](https://github.com/arthurgeron/eslint-plugin-react-usememo/blob/main/docs/rules/require-usememo.md#options).

**Comment by arthurgeron-work** [comment]
On second thought, I think `memo` may still be useful even for components that receives no props, as it'll prevent re-renders from parent components from causing re-renders in the child component wrapped with memo!

**Comment by arthurgeron-work** [comment]
> however, there is a case when it does not work correctly:
> a simple function with an uppercase name and parameters:
> export function Test(params: any) {} or export const Test = (params: any) => {}

Pleaset correct me if I didn't understand your issue, but functions in PascalCase should always be components in React projects, no exceptions! 
The generally agreed naming conventions for React is that:
- PascalCase is used for naming components and higher-order components.
- camelCase is used for naming instances, variables, and functions that are not components.

**Comment by arthurgeron-work** [comment]
@veksa  I've added `ignoredComponents` options to the `require-memo` rule, you can check more info on it [here](https://github.com/arthurgeron/eslint-plugin-react-usememo/blob/05f3524becb752fe83edd716014010a0e354efd3/docs/rules/require-memo.md#options).

It's been published to v `2.3.0--beta4`, I'll publish it to release later today.

Let me know if there's anything else.
